### ISSUE TITLE
v2.1.0

### ISSUE THREAD

**Issue by andris9** [nodemailer/nodemailer#549]
These are the planned changes for Nodemailer v2.1.0. There are going to be some new features (most probably templating support) but it is going to be backwards compatible, so it bumps only minor version.

### Message  headers

  * [x] Make sure custom headers work the same in every cases (message root, attachments and alternatives) (*v2.1.0-beta.0*)
  * [x] Array as an header value should be converted into several header rows with the same key (*v2.1.0-beta.0*)

```javascript
var headers = {
    'X-My-Header': ['foo', 'bar']
}
/*
X-My-Header: foo
X-My-Header: bar
*/
```
  * [x] Allow unprocessed headers (do not encode, do not fold etc.). (*v2.1.0-beta.1*)

```javascript
var headers = {
    'X-My-Header': {
        prepared: true,
        value: 'do not fold or encode this value, pass it to message header as is'
    }
}
```
### Misc

  * [x] Message-Id should be auto generated immediatelly after `sendMail` is called and not wait until message starts streaming (*v2.1.0-beta.0*)

### Templating

I still want to bring templating support to Nodemailer.

  * [x] Add support for templates (*v2.1.0-beta.2*, details [here](https://github.com/nodemailer/nodemailer/blob/aa3b87ae2489b61d373d208714e8382985648831/README.md#using-templates), might change)

```javascript
// create template based sender function
var sendPwdReminder = transporter.templateSender({
    // template fields
    subject: 'Password reminder for {{username}}!',
    text: 'Hello, {{username}}, Your password is: {{ password }}',
    html: '<b>Hello, <strong>{{username}}</strong>, Your password is:\n<b>{{ password }}</b></p>'
}, {
    // normal fields added to every message
    from: 'sender@example.com',
});

// use template based sender to send a message
sendPwdReminder({
    // normal fields added to this message
    to: 'receiver@example.com'
}, {
    // context values for template fields
    username: 'Node Mailer',
    password: '!"\'<>&some-thing'
}, function(err, info){
    if(err){
       console.log('Error');
    }else{
        console.log('Password reminder sent');
    }
});
```
  * [x] Add support for custom renderers. This would probably mean supporting [node-email-templates](https://github.com/niftylettuce/node-email-templates) template objects (or actually any object with `render(context, cb)` method). (*v2.1.0-beta.3*)

```javascript
var sendPwdReminder = transporter.templateSender(new EmailTemplate(templateDir));
```

**Comment by andris9** [comment]
Released v2.1.0-beta.0, install with ```npm install nodemailer@beta```

**Comment by andris9** [comment]
Released v2.1.0-beta.1. Added `prepared` flag to custom headers to prevent processing of this header value

```javascript
var headers = {
    'X-My-Header': {
        prepared: true,
        value: 'do not fold or encode this value, pass it to message header as is'
    }
}
```

**Comment by andris9** [comment]
Released v2.1.0-beta.2. Implemented simple templating system. See details [here](https://github.com/nodemailer/nodemailer/blob/aa3b87ae2489b61d373d208714e8382985648831/README.md#using-templates)

**Comment by ukriseq** [comment]
Hi,

Thanks for the awesome plugin. When sending content images inside html, the mail is rendering fine with the embedded images  - but the mail is being shown as having attachments - when there are no attachments being sent - only content related images - Can u please fix this

**Comment by andris9** [comment]
@ukriseq could you create a separate issue for this. Also provide an example script that generates such messages

**Comment by andris9** [comment]
Released v2.1.0-rc.0 install with `npm install nodemailer@beta`. [full docs](https://github.com/nodemailer/nodemailer/blob/v2.1/README.md).

This is a rc release, so if there will be no changes from now then the stable will be released on january 27th (stable is released 7 days after latest rc release).

**Comment by andris9** [comment]
Released v2.1.0-rc.1 where I sneaked in a few changes even though it is a rc.

  * If a SMTP pool is closed and it still has some pending messages in queue, these callbackes are returned with an error (so far these messages were kept hanging indefinitely)
  * tranporter object emits an 'idle' event for smtp pool if there are free connection slots available. Check if the tranporter is still idling with `transporter.isIdle()`

```javascript
transporter.on('idle', function(){
    var message;
    while(transport.isIdle() && (message = retrieveNext())){
        transporter.sendMail(message);
    }
});
```

As updates were introduced, then stable release date is moved from january 27th to february 1st.
